core/txpool/bundlepool/bundlepool.go

```diff
@@ -0,0 +1,456 @@
+
+	maxMinTimestampFromNow = int64(300) // 5 minutes
+
+	dialAttempts = 3
+	dialSleep    = 3 * time.Second
+)
+
+var (
+	bundleGauge = metrics.NewRegisteredGauge("bundlepool/bundles", nil)
+	slotsGauge  = metrics.NewRegisteredGauge("bundlepool/slots", nil)
+)
+
+var (
+	// ErrSimulatorMissing is returned if the bundle simulator is missing.
+	ErrSimulatorMissing = errors.New("bundle simulator is missing")
+
+	// ErrBundleTimestampTooHigh is returned if the bundle's MinTimestamp is too high.
+	ErrBundleTimestampTooHigh = errors.New("bundle MinTimestamp is too high")
+
+	// ErrBundleGasPriceLow is returned if the bundle gas price is too low.
+	ErrBundleGasPriceLow = errors.New("bundle gas price is too low")
+
+	// ErrBundleAlreadyExist is returned if the bundle is already contained
+	// within the pool.
+	ErrBundleAlreadyExist = errors.New("bundle already exist")
+)
+
+// BlockChain defines the minimal set of methods needed to back a tx pool with
+// a chain. Exists to allow mocking the live chain out of tests.
+type BlockChain interface {
+	// Config retrieves the chain's fork configuration.
+	Config() *params.ChainConfig
+
+	// CurrentBlock returns the current head of the chain.
+	CurrentBlock() *types.Header
+
+	// GetBlock retrieves a specific block, used during pool resets.
+	GetBlock(hash common.Hash, number uint64) *types.Block
+
+	// StateAt returns a state database for a given root hash (generally the head).
+	StateAt(root common.Hash) (*state.StateDB, error)
+}
+
+type BundleSimulator interface {
+	SimulateBundle(bundle *types.Bundle) (*big.Int, error)
+}
+
+type BundlePool struct {
+	config    Config
+	mevConfig miner.MevConfig
+
+	bundles    map[common.Hash]*types.Bundle
+	bundleHeap BundleHeap
+	mu         sync.RWMutex
+
+	slots uint64 // Number of slots currently allocated
+
+	simulator BundleSimulator
+
+	bundleReceiverClients map[string]*rpc.Client
+	deliverBundleCh       chan *types.SendBundleArgs
+	exitCh                chan struct{}
+}
+
+func New(config Config, mevConfig miner.MevConfig) *BundlePool {
+	// Sanitize the input to ensure no vulnerable gas prices are set
+	config = (&config).sanitize()
+
+	pool := &BundlePool{
+		config:                config,
+		mevConfig:             mevConfig,
+		bundles:               make(map[common.Hash]*types.Bundle),
+		bundleHeap:            make(BundleHeap, 0),
+		bundleReceiverClients: make(map[string]*rpc.Client),
+		deliverBundleCh:       make(chan *types.SendBundleArgs),
+		exitCh:                make(chan struct{}),
+	}
+	if pool.mevConfig.MevReceivers == nil {
+		return pool
+	}
+	for _, v := range mevConfig.MevReceivers {
+		pool.register(v)
+	}
+	if len(pool.bundleReceiverClients) == 0 {
+		log.Error("No valid bundleReceivers")
+	}
+	go pool.deliverLoop()
+
+	return pool
+}
+
+func (p *BundlePool) deliverLoop() {
+	for {
+		select {
+		case bundle := <-p.deliverBundleCh:
+			for url, cli := range p.bundleReceiverClients {
+				cli := cli
+				url := url
+				go func() {
```

Comment:
I didn't get the design here. If we are going to send the bundles in goroutines anyway, why do we bother use `deliverBundleCh`?
Why don't we just have a `deliverBundle` function and use a separate goroutine to call it when we want to deliver the bundle?

If we want to control the concurrency, we should use a buffered channel and a reasonable number of consumers to do the delivery.

Reply:
There is indeed unnecessary to use channel to implement, which increases complexity. Now it is simplified, please review again.